src/main/java/com/ashkan/moviesapi/services/implementations/MovieRentalServiceImpl.java

```diff
@@ -0,0 +1,109 @@
+package com.ashkan.moviesapi.services.implementations;
+
+import com.ashkan.moviesapi.constants.constants;
+import com.ashkan.moviesapi.repositories.MemberRepository;
+import com.ashkan.moviesapi.repositories.MovieCatalogRepository;
+import com.ashkan.moviesapi.repositories.MovieRentalRepository;
+import com.ashkan.moviesapi.repositories.MovieRepository;
+import com.ashkan.moviesapi.entities.Member;
+import com.ashkan.moviesapi.entities.Movie;
+import com.ashkan.moviesapi.entities.MovieCatalog;
+import com.ashkan.moviesapi.entities.MovieRental;
+import com.ashkan.moviesapi.exceptions.BadRequest.MovieRentalBadRequestException;
+import com.ashkan.moviesapi.exceptions.BadRequest.MovieRentalBadRequestException_NoCopies;
+import com.ashkan.moviesapi.exceptions.NotFound.MovieRentalNotFoundException;
+import com.ashkan.moviesapi.services.interfaces.MovieRentalService;
+import org.modelmapper.ModelMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.sql.Timestamp;
+import java.util.List;
+
+@Service
+public class MovieRentalServiceImpl implements MovieRentalService {
+    private MovieRentalRepository movieRentalRepository;
+    private MovieRepository movieRepository;
+    private MemberRepository memberRepository;
+    private MovieCatalogRepository movieCatalogRepository;
+
+    @Autowired
+    public MovieRentalServiceImpl(MovieRentalRepository movieRentalRepository,
+                                  MovieRepository movieRepository,
+                                  MemberRepository memberRepository,
+                                  MovieCatalogRepository movieCatalogRepository) {
+        this.movieRentalRepository = movieRentalRepository;
+        this.movieRepository = movieRepository;
+        this.memberRepository = memberRepository;
+        this.movieCatalogRepository = movieCatalogRepository;
+    }
+
+    @Override
+    public List<MovieRental> findAll() {
+        return movieRentalRepository.findAll();
+    }
+
+    @Override
+    public MovieRental findById(int id) {
+        return movieRentalRepository.findById(id)
+                .orElseThrow(() -> new MovieRentalNotFoundException(id));
+    }
+
+    @Override
+    public void save(MovieRental movieRental) {
+        Movie movie = movieRepository.findById(movieRental.getMovieId())
+                .orElseThrow(() -> new MovieRentalNotFoundException(movieRental.getMovieId()));
+
+        Member member = memberRepository.findById(movieRental.getMemberId())
+                .orElseThrow(() -> new MovieRentalNotFoundException(movieRental.getMemberId()));
+
+        MovieCatalog movieCatalog = movieCatalogRepository.findById(movieRental.getMovieId())
+                .orElseThrow(() -> new RuntimeException("Movie catalog not found " + movieRental.getMovieId()));
+
+        Timestamp currentDate = new Timestamp(System.currentTimeMillis());
```

Comment:
According to the requirements the rentalDate should be sent in the payload, you should not generate this value in the API, then you need to add a validation to check that this date is not a past date.